File: TOP100/200.java
class Solution {

  int m;
  int n;
  int count = 0;

  public int numIslands(char[][] grid) {
    m = grid.length;
    n = grid[0].length;

    for (int i = 0; i < m; i++) {
      check(grid, i);
    }

    return count;
  }

  public void check(char[][] grid, int i) {
    for (int j = 0; j < n; j++) {
      if (grid[i][j] == '1') {
        bfs(grid, i, j);
        count++;
      }
    }
  }

  public void bfs(char[][] grid, int i, int j) {
    grid[i][j] = '2';

    if (i - 1 >= 0 && grid[i - 1][j] == '1') bfs(grid, i - 1, j);

    if (i + 1 < m && grid[i + 1][j] == '1') bfs(grid, i + 1, j);

    if (j - 1 >= 0 && grid[i][j - 1] == '1') bfs(grid, i, j - 1);

    if (j + 1 < n && grid[i][j + 1] == '1') bfs(grid, i, j + 1);
  }
}

File: 2024_april/200.js
var numIslands = function(grid) {
  
    var count = 0;
    var n = grid.length;
    var m = grid[0].length;

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (grid[i][j] == '1') {
                recursive(grid, i, j, n, m);
                count++;
            }
        }
    }

    return count;
};

function recursive(grid, i, j, n , m) {
    grid[i][j] = '2';

    if (i - 1 >= 0 && grid[i - 1][j] == '1') {
        recursive(grid, i - 1, j, n, m);
    }

    if (i + 1 < n && grid[i + 1][j] == '1') {
        recursive(grid, i + 1, j, n, m);
    }

    if (j - 1 >= 0 && grid[i][j - 1] == '1') {
        recursive(grid, i, j - 1, n, m);
    }
    
    if (j + 1 < m && grid[i][j + 1] == '1') {
        recursive(grid, i, j + 1, n, m);
    }
}